packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js

```diff
@@ -0,0 +1,239 @@
+    onItemsFetch: OnItemsFetchFunction;
+    onScroll: (o: ScrollEvent) => void;
+    sentinel: ?HTMLElement;
+
+    static propTypes = {
+      elementsPerScroll: PropTypes.number,
+      loadingComponent: PropTypes.node,
+      onItemsFetch: PropTypes.func.isRequired,
+      onClickSeeMore: PropTypes.func,
+      onScroll: PropTypes.func,
+      onScrollFinished: PropTypes.func,
+      seeMoreComponent: PropTypes.node,
+      seeMoreAfter: PropTypes.number,
+    };
+
+    static defaultProps = {
+      elementsPerScroll: 5,
+      loadingComponent: null,
+      onClickSeeMore: () => {},
+      onScroll: () => {},
+      onScrollFinished: () => {},
+      seeMoreComponent: null,
+      seeMoreAfter: null,
+    };
+
+    constructor(props: Props) {
+      super(props);
+
+      this.state = {
+        index: 0,
+        elementsToRender: [],
+        isListFinished: false,
+        showSeeMore: false,
+      };
+
+      this.onItemsFetch = props.onItemsFetch;
+      this.onScroll = props.onScroll;
+      this.observer = new IntersectionObserver(this.handleIntersection, {
+        threshold: 1,
+      });
+    }
+
+    componentDidMount() {
+      if (this.sentinel) {
+        this.observer.observe(this.sentinel);
+      }
+    }
+
+    componentDidUpdate() {
+      if (this.sentinel && this.state.index > 0) {
+        this.observer.observe(this.sentinel);
+      }
+    }
+
+    loadMore() {
+      return new Promise(resolve => {
+        const { index, elementsToRender } = this.state;
+        const { elementsPerScroll, seeMoreAfter } = this.props;
+        this.onItemsFetch(index, elementsPerScroll).then(nextElements => {
+          if (nextElements && nextElements.length > 0) {
+            const nextIndex = index + elementsPerScroll;
+            return resolve({
+              index: nextIndex,
+              elementsToRender: elementsToRender.concat(nextElements),
+              showSeeMore: seeMoreAfter === index / elementsPerScroll,
+            });
+          }
+          this.props.onScrollFinished({
+            totalNumberElements: elementsToRender.length,
+          });
+          return resolve({
+            isListFinished: true,
+          });
+        });
+      });
+    }
+
+    handleIntersection = (entries: Array<IntersectionObserverEntry>) =>
+      new Promise(resolve => {
+        const entry = entries[0];
+        if (entry.intersectionRatio === 1) {
+          if (this.sentinel) {
+            this.observer.unobserve(this.sentinel);
+          }
+          this.props.onScroll({ currentIndex: this.state.index });
+          this.loadMore().then(newState => {
+            this.setState(newState);
+            resolve();
+          });
+        } else {
+          resolve();
+        }
+      });
+
+    handleKeyPress = (e: SyntheticKeyboardEvent<HTMLButtonElement>) => {
+      const { key } = e;
+      if (key === 'Enter' || key === ' ') {
+        return this.handleSeeMoreClick(e);
+      }
+      return null;
```

Comment:
change this to `return Promise.resolve()` is never a good idea to have a function that sometimes returns a promise and sometimes not.

Reply:
done